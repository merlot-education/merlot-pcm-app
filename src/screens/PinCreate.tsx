import React, { useState, useEffect } from 'react'
import { Alert, Keyboard, StyleSheet, View } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import ReactNativeBiometrics from 'react-native-biometrics'
import { useTranslation } from 'react-i18next'
import md5 from 'md5'
import { getValueKeychain, setValueKeychain } from '../utils/keychain'
import { Colors } from '../theme/theme'
import { Loader, TextInput } from '../components'
import Button, { ButtonType } from '../components/button/Button'
import * as api from '../api'

interface PinCreateProps {
  setAuthenticated: React.Dispatch<React.SetStateAction<boolean>>
  initAgent: (email: string, pin: string) => void
  route
}

const style = StyleSheet.create({
  container: {
    backgroundColor: Colors.background,
    margin: 20,
  },
  btnContainer: {
    marginTop: 20,
  },
})

const PinCreate: React.FC<PinCreateProps> = ({
  setAuthenticated,
  initAgent,
  route,
}) => {
  const [pin, setPin] = useState('')
  const [pinTwo, setPinTwo] = useState('')
  const [biometricSensorAvailable, setBiometricSensorAvailable] =
    useState(false)
  const [successPin, setSuccessPin] = useState(false)
  const [successBiometric, setSuccessBiometric] = useState(false)
  const [loading, setLoading] = useState(false)
  const [email, setEmail] = useState('')
  const { forgotPin } = route.params
  const { t } = useTranslation()

  const sendSeedHash = async (userEmail: string) => {
    const genratedSeedHash = md5(userEmail)
    const seedHashResponse = await api.default.auth.sendSeedHash({
      email: userEmail,
      seedHash: genratedSeedHash,
    })
    if (seedHashResponse.data != null) {
      // seed sent via email
    }
  }
  const startAgent = async (email: string, pin: string) => {
    setLoading(true)
    await initAgent(email, pin)
    await sendSeedHash(email)
    // setTimeout(() => {
    setLoading(false)
    setAuthenticated(true)
    // }, 10000)
  }
  useEffect(() => {
    ReactNativeBiometrics.isSensorAvailable().then(resultObject => {
      const { available, biometryType } = resultObject
      if (available && biometryType === ReactNativeBiometrics.Biometrics) {
        setBiometricSensorAvailable(true)
      }
    })
  })

  const passcodeCreate = async (passcode: string) => {
    const description = t('PinCreate.UserAuthenticationPin')
    try {
      setValueKeychain(description, passcode, {
        service: 'passcode',
      })

      // Change email here
      const keychainEntry = await getValueKeychain({
        service: 'email',
      })
      setEmail(keychainEntry.password)
      Alert.alert(t('PinCreate.PinsSuccess'), '', [
        {
          text: 'Ok',
          // onPress: () =>
          //   startAgent(JSON.parse(keychainEntry).password, passcode),
        },
      ])
      setSuccessPin(true)
    } catch (e) {
      Alert.alert(e)
    }
  }

  const confirmEntry = (x: string, y: string) => {
    if (x.length < 6 || y.length < 6) {
      Alert.alert(t('PinCreate.PinMustBe6DigitsInLength'))
    } else if (x !== y) {
      Alert.alert(t('PinCreate.PinsEnteredDoNotMatch'))
    } else {
      passcodeCreate(x)
    }
  }
  const biometricEnable = () => {
    ReactNativeBiometrics.isSensorAvailable().then(resultObject => {
      const { available, biometryType } = resultObject
      if (available && biometryType === ReactNativeBiometrics.Biometrics) {
        ReactNativeBiometrics.simplePrompt({
          promptMessage: t('Biometric.BiometricConfirm'),
        })
          .then(resultObject => {
            const { success } = resultObject

            if (success) {
              ReactNativeBiometrics.createKeys().then(() => {
                setSuccessBiometric(true)
                Alert.alert(t('Biometric.BiometricSuccess'))
              })
            } else {
              Alert.alert(t('Biometric.BiometricCancle'))
            }
          })
          .catch(() => {
            Alert.alert(t('Biometric.BiometricFailed'))
          })
      } else {
        Alert.alert(t('Biometric.BiometricNotSupport'))
      }
    })
  }

  const onSubmit = async () => {
    if (successPin && successBiometric) {
      // setAuthenticated(true)
      await startAgent(email, pin)
    } else if (successPin && !biometricSensorAvailable) {
      setAuthenticated(true)
    } else {
      Alert.alert(t('Biometric.RegisterPinandBiometric'))
    }
  }
  return (
    <SafeAreaView style={[style.container]}>
      <Loader loading={loading} />
      <TextInput
        label={t('Global.EnterPin')}
        placeholder={t('Global.6DigitPin')}
        placeholderTextColor={Colors.lightGrey}
        accessible
        accessibilityLabel={t('Global.EnterPin')}
        maxLength={6}
        autoFocus
        secureTextEntry
        keyboardType="number-pad"
        value={pin}
        onChangeText={setPin}
      />
      <TextInput
        label={t('PinCreate.ReenterPin')}
        accessible
        accessibilityLabel={t('PinCreate.ReenterPin')}
        placeholder={t('Global.6DigitPin')}
        placeholderTextColor={Colors.lightGrey}
        maxLength={6}
        secureTextEntry
        keyboardType="number-pad"
        value={pinTwo}
        onChangeText={(text: string) => {
          setPinTwo(text)
          if (text.length === 6) {
            Keyboard.dismiss()
          }
        }}
      />
      <Button
        title="Setup PIN"
        buttonType={ButtonType.Primary}
        onPress={() => {
          Keyboard.dismiss()
          confirmEntry(pin, pinTwo)
        }}
      />
      {!forgotPin && (
        <>
          <View style={style.btnContainer}>
            {biometricSensorAvailable && (
              <Button
                title="Setup Biometric"
                buttonType={ButtonType.Primary}
                onPress={biometricEnable}
              />
            )}
          </View>
          <View style={style.btnContainer}>
            <Button
              title="Create Wallet"
              buttonType={ButtonType.Primary}
              onPress={onSubmit}
            />
          </View>
        </>
      )}
    </SafeAreaView>
  )
}

export default PinCreate
