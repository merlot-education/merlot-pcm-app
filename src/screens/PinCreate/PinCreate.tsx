import React, { useState, useEffect, useCallback } from 'react'
import {
  Keyboard,
  StyleSheet,
  Text,
  View,
  Alert,
  BackHandler,
} from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import ReactNativeBiometrics from 'react-native-biometrics'
import { useTranslation } from 'react-i18next'
import { StackScreenProps } from '@react-navigation/stack'

import { ColorPallet, TextTheme } from '../../theme/theme'
import { Loader, TextInput } from '../../components'
import Button, { ButtonType } from '../../components/button/Button'
import { KeychainStorageKeys } from '../../constants'
import { OnboardingStackParams, Screens } from '../../types/navigators'
import {
  checkIfSensorAvailable,
  createBiometricKeys,
  createMD5HashFromString,
  getValueFromKeychain,
  saveValueInKeychain,
  showBiometricPrompt,
  storeOnboardingCompleteStage,
} from './PinCreate.utils'
import { errorToast, successToast, warningToast } from '../../utils/toast'
import { getValueKeychain } from '../../utils/keychain'

type PinCreateProps = StackScreenProps<OnboardingStackParams, Screens.CreatePin>

const style = StyleSheet.create({
  container: {
    backgroundColor: ColorPallet.grayscale.white,
    margin: 20,
    flex: 1,
  },
  btnContainer: {
    marginTop: 20,
  },
  label: {
    ...TextTheme.label,
    fontWeight: 'bold',
    textAlign: 'center',
  },
})

const PinCreate: React.FC<PinCreateProps> = ({ navigation, route }) => {
  const { forgotPin } = route.params
  const [pin, setPin] = useState('')
  const [pinTwo, setPinTwo] = useState('')
  const [biometricSensorAvailable, setBiometricSensorAvailable] =
    useState(false)
  const [successPin, setSuccessPin] = useState(false)
  const [successBiometric, setSuccessBiometric] = useState(false)
  const [loading, setLoading] = useState(false)
  const { t } = useTranslation()

  const checkBiometricIfPresent = useCallback(async () => {
    const { available, biometryType } = await checkIfSensorAvailable()
    if (available && biometryType === ReactNativeBiometrics.Biometrics) {
      setBiometricSensorAvailable(true)
    }
  }, [])

  useEffect(() => {
    checkBiometricIfPresent()
  }, [checkBiometricIfPresent])

  useEffect(() => {
    const backAction = () => {
      Alert.alert(
        'Already authenticated!',
        'Are you sure you want to go back?',
        [
          {
            text: 'Cancel',
            onPress: () => null,
            style: 'cancel',
          },
          {
            text: 'YES',
            onPress: () =>
              navigation.navigate(Screens.Registration, { forgotPin: false }),
          },
        ],
      )
      return true
    }

    const backActionForgotPassword = () => {
      navigation.navigate(Screens.EnterPin)
      return true
    }
    const backHandler = BackHandler.addEventListener(
      'hardwareBackPress',
      forgotPin ? backActionForgotPassword : backAction,
    )

    return () => backHandler.remove()
  }, [forgotPin, navigation])

  const passcodeCreate = async (passcode: string) => {
    try {
      const [email, passphrase] = await Promise.all([
        new Promise(resolve => {
          resolve(getValueFromKeychain(KeychainStorageKeys.Email))
        }),
        new Promise(resolve => {
          resolve(getValueFromKeychain(KeychainStorageKeys.Passphrase))
        }),
        new Promise(resolve => {
          resolve(
            saveValueInKeychain(
              KeychainStorageKeys.Passcode,
              passcode,
              t('PinCreate.UserAuthenticationPin'),
            ),
          )
        }),
      ])
      setSuccessPin(true)
      if (forgotPin) {
        navigation.navigate(Screens.EnterPin)
      }
      successToast(t('PinCreate.PinsSuccess'))
    } catch (e) {
      errorToast(e)
    }
  }

  const confirmEntry = async (pin: string, reEnterPin: string) => {
    if (pin.length < 6) {
      warningToast(t('PinCreate.PinMustBe6DigitsInLength'))
    } else if (reEnterPin.length < 6) {
      errorToast(t('PinCreate.ReEnterPinMustBe6DigitsInLength'))
    } else if (pin !== reEnterPin) {
      errorToast(t('PinCreate.PinsEnteredDoNotMatch'))
    } else {
      await passcodeCreate(pin)
    }
  }

  const biometricEnable = async () => {
    const { available, biometryType } = await checkIfSensorAvailable()
    if (available && biometryType === ReactNativeBiometrics.Biometrics) {
      const { success, error } = await showBiometricPrompt()
      if (success) {
        await createBiometricKeys()
        setSuccessBiometric(true)
        successToast(t('Biometric.BiometricSuccess'))
      } else {
        warningToast(error)
      }
    } else {
      warningToast(t('Biometric.BiometricNotSupport'))
    }
  }

  const onSubmit = async () => {
    if (successPin && successBiometric) {
      navigation.navigate(Screens.CreateWallet)
    } else if (successPin && !biometricSensorAvailable) {
      navigation.navigate(Screens.CreateWallet)
    } else {
      warningToast(t('PinCreate.RegisterPinandBiometric'))
    }
  }
  const showSameEmailAlert = () => {
    Alert.alert(t('PinCreate.EmailConfirmation'), t('PinCreate.CheckEmail'), [
      {
        text: t('Global.ChangeEmail'),
        style: 'cancel',
        onPress: () =>
          navigation.navigate(Screens.Registration, { forgotPin: false }),
      },
      { text: t('Global.Next'), onPress: proceedToImport },
    ])
  }

  const proceedToImport = () => {
    if (successPin && successBiometric) {
      navigation.navigate(Screens.ImportWallet)
    } else if (successPin && !biometricSensorAvailable) {
      navigation.navigate(Screens.ImportWallet)
    } else {
      warningToast(t('PinCreate.RegisterPinandBiometric'))
    }
  }

  const onImportWallet = () => {
    showSameEmailAlert()
  }

  return (
    <SafeAreaView style={[style.container]}>
      <Loader loading={loading} />
      <TextInput
        label={t('Global.EnterPin')}
        placeholder={t('Global.6DigitPin')}
        placeholderTextColor={ColorPallet.baseColors.lightGrey}
        accessible
        accessibilityLabel={t('Global.EnterPin')}
        maxLength={6}
        autoFocus
        secureTextEntry
        keyboardType="number-pad"
        value={pin}
        onChangeText={setPin}
      />
      <TextInput
        label={t('PinCreate.ReenterPin')}
        accessible
        accessibilityLabel={t('PinCreate.ReenterPin')}
        placeholder={t('Global.6DigitPin')}
        placeholderTextColor={ColorPallet.baseColors.lightGrey}
        maxLength={6}
        secureTextEntry
        keyboardType="number-pad"
        value={pinTwo}
        onChangeText={(text: string) => {
          setPinTwo(text)
          if (text.length === 6) {
            Keyboard.dismiss()
          }
        }}
        editable={pin.length === 6 && true}
      />
      <View style={style.btnContainer}>
        <Button
          title="Setup PIN"
          buttonType={ButtonType.Primary}
          disabled={successPin}
          onPress={() => {
            Keyboard.dismiss()
            confirmEntry(pin, pinTwo)
          }}
        />
      </View>
      {!forgotPin && (
        <>
          <View style={style.btnContainer}>
            {biometricSensorAvailable && (
              <Button
                title="Setup Biometric"
                buttonType={ButtonType.Primary}
                onPress={biometricEnable}
                disabled={successBiometric}
              />
            )}
          </View>
          <View style={style.btnContainer}>
            <Button
              title="Import Wallet"
              buttonType={ButtonType.Primary}
              onPress={onImportWallet}
            />
          </View>
          <View style={style.btnContainer}>
            <Text style={style.label}> {t('PinCreate.OR')}</Text>
          </View>
          <View style={style.btnContainer}>
            <Button
              title="Create Wallet"
              buttonType={ButtonType.Primary}
              onPress={onSubmit}
            />
          </View>
        </>
      )}
    </SafeAreaView>
  )
}

export default PinCreate
