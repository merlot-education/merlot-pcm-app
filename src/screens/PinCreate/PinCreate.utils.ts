import AsyncStorage from '@react-native-async-storage/async-storage';
import { Alert } from 'react-native';
import ReactNativeBiometrics from 'react-native-biometrics';
import md5 from 'md5';
import i18next from 'i18next';
import { LocalStorageKeys } from '../../constants';
import { getValueKeychain, setValueKeychain } from '../../utils/keychain';

const rnBiometrics = new ReactNativeBiometrics();

export const storeOnboardingCompleteStage = async () => {
  await AsyncStorage.setItem(LocalStorageKeys.OnboardingCompleteStage, 'true');
};

export const getValueFromKeychain = async (key: string) => {
  const data = await getValueKeychain({
    service: key,
  });
  return data;
};

export const saveValueInKeychain = async (
  service: string,
  value: string,
  description: string,
) => {
  try {
    await setValueKeychain(description, value, {
      service,
    });
  } catch (e: any) {
    Alert.alert(e);
  }
};

export const checkIfSensorAvailable = async () => {
  const result = await rnBiometrics.isSensorAvailable();
  return result;
};

export const showBiometricPrompt = async () => {
  const result = await rnBiometrics.simplePrompt({
    promptMessage: i18next.t('Biometric.BiometricConfirm'),
  });
  return result;
};

export const createBiometricKeys = async () => {
  const result = await rnBiometrics.createKeys();
  return result;
};

export const createMD5HashFromString = (value: string) => {
  const hash = String(md5(value));
  return hash;
};
