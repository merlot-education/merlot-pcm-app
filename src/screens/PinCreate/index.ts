import PinCreate from './PinCreate'

export default PinCreate
