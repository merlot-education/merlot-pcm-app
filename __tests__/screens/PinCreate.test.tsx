import React from 'react'
import { create } from 'react-test-renderer'
import PinCreate from '../../src/screens/PinCreate'

jest.mock('react-native-keychain', () => 'PinCreate')
jest.mock('react-native-config', () => 'PinCreate')

describe('PinCreate Screen', () => {
  it('PinCreate correctly', () => {
    const tree = create(<PinCreate />).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
