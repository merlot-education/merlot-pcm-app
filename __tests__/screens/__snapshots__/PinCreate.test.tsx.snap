// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PinCreate Screen PinCreate correctly 1`] = `
<RNCSafeAreaView
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "margin": 20,
      },
    ]
  }
>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(52, 52, 52, 0.8)",
          "flex": 1,
          "justifyContent": "space-around",
        }
      }
    >
      <ActivityIndicator
        color="#03028A"
        size="small"
      />
    </View>
  </Modal>
  <View
    style={
      Object {
        "marginVertical": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#03028A",
          "fontSize": 14,
          "fontWeight": "bold",
          "marginBottom": 3,
        }
      }
    >
      Global.EnterPin
    </Text>
    <TextInput
      accessibilityLabel="Global.EnterPin"
      accessible={true}
      autoFocus={true}
      keyboardType="number-pad"
      maxLength={6}
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="Global.6DigitPin"
      placeholderTextColor="#D3D3D3"
      secureTextEntry={true}
      selectionColor="#03028A"
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": undefined,
            "borderRadius": 4,
            "borderWidth": 2,
            "color": "#03028A",
            "fontSize": 16,
            "padding": 10,
          },
          false,
        ]
      }
      value=""
    />
  </View>
  <View
    style={
      Object {
        "marginVertical": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#03028A",
          "fontSize": 14,
          "fontWeight": "bold",
          "marginBottom": 3,
        }
      }
    >
      PinCreate.ReenterPin
    </Text>
    <TextInput
      accessibilityLabel="PinCreate.ReenterPin"
      accessible={true}
      keyboardType="number-pad"
      maxLength={6}
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="Global.6DigitPin"
      placeholderTextColor="#D3D3D3"
      secureTextEntry={true}
      selectionColor="#03028A"
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": undefined,
            "borderRadius": 4,
            "borderWidth": 2,
            "color": "#03028A",
            "fontSize": 16,
            "padding": 10,
          },
          false,
        ]
      }
      value=""
    />
  </View>
  <View
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 5,
          "borderWidth": 1,
          "justifyContent": "center",
        },
        Object {
          "height": 47,
          "paddingLeft": 15,
          "paddingRight": 15,
        },
        Object {
          "backgroundColor": "#108ee9",
          "borderColor": "#108ee9",
        },
        false,
        false,
        undefined,
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 18,
            },
            Object {
              "color": "#ffffff",
            },
            false,
            false,
          ]
        }
      >
        Setup PIN
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "marginTop": 20,
      }
    }
  />
  <View
    style={
      Object {
        "marginTop": 20,
      }
    }
  >
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 5,
            "borderWidth": 1,
            "justifyContent": "center",
          },
          Object {
            "height": 47,
            "paddingLeft": 15,
            "paddingRight": 15,
          },
          Object {
            "backgroundColor": "#108ee9",
            "borderColor": "#108ee9",
          },
          false,
          false,
          undefined,
          undefined,
        ]
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 18,
              },
              Object {
                "color": "#ffffff",
              },
              false,
              false,
            ]
          }
        >
          Create Wallet
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
